### ISSUE TITLE
MemoryBuffer::create_from_memory_range is unsound

### ISSUE THREAD

**Issue by programmerjake** [TheDan64/inkwell#241]
the `input` slice is borrowed by the returned `MemoryBuffer`, so `create_from_memory_range` needs to be `unsafe` and/or the returned `MemoryBuffer` type needs to have the same lifetime as `input`.

So, something like:
```rust
pub struct MemoryBuffer<'a> {
    value: LLVMMemoryBuffer, // or whatever
    _phantom: PhantomData<&'a [u8]>,
}

impl MemoryBuffer<'a> {
    pub fn create_from_memory_range(input: &'a [u8], name: &str) -> MemoryBuffer<'a> {
        // Safety:
        // `input` is borrowed by the new memory buffer, name is copied,
        // null-termination of `input` is not required since we passed `false` there.
        unsafe {
            let name = CString::new(name).unwrap();
            let value = LLVMCreateMemoryBufferWithMemoryRange(
                input.as_ptr() as *const _,
                input.len(),
                name.as_ptr(),
                false as _);
        }
        MemoryBuffer { value, _phantom: PhantomData }
    }
}
```

See also #185

**Comment by TheDan64** [comment]
I think `MemoryBuffer` can sometimes own its data, so we'd need a ref and owning type, I think

**Comment by programmerjake** [comment]
for the owning type, just use `MemoryBuffer<'static>`

**Comment by TheDan64** [comment]
I thought about that too, and it's a valid option. But it feels a little dirty since there's nothing actually happening involving a `'static` lifetime

**Comment by programmerjake** [comment]
Well, IMHO, `'static` just means you don't have to worry about the type holding references to things that might be left dangling -- exactly what you want. Just like a `Cow<'static, str>` can be `Cow::Owned`.

**Comment by TheDan64** [comment]
True

**Comment by Anton-4** [comment]
This issue has a high severity rating on cargo-crev, could you add a fitting label to this issue?

**Comment by TheDan64** [comment]
Sure thing. Also, I agree with the proposed solution and would likely accept a PR if someone has the time

**Comment by Anton-4** [comment]
Thanks :)
I've put it on my to do list but it may take a while until I get to it.

**Comment by ErisianArchitect** [comment]
@TheDan64 This issue also appears to have been solved by #654 .

**Comment by programmerjake** [comment]
yeah, this has been fixed, with the follow-on fix in #655 after #654 improperly restricted the contents to not contain inner nul bytes (e.g. for object files) by using `CStr`.
